### DISCUSSION TITLE
How to format "epoch" in MessageFormatter ?

### DISCUSSION THREAD

**Discussion by jmbeuken** [buildbot/buildbot#6474] [Q&A] [answered]
Hi,

```
message_formatter_inst = MessageFormatter(
   template = u'''\
{% set Epoch = buildset['submitted_at'] %}
<p><b>Epoch</b> : {{ Epoch }}</p>
   ''',
   ...
)
```

**Epoch** : 1648888349

How to "format" Epoch in template to obtain readable date in message ?  

Like :    datetime.fromtimestamp(**Epoch**).strftime('%c')   ->    Sun Mar 27 18:33:45 2022

Need to subclass MessageFormatter ?

Thank you,
Jean-Michel

**Comment by vibbo** [comment]
Hi, you can enhance the `MessageFormatter` via `ctx` parameter.
See `ctx` in https://docs.buildbot.net/latest/manual/configuration/report_generators/formatter.html#messageformatter.
And look at the example.

Hope this helps ;)

**Comment by jmbeuken** [comment]
Hi,
Thx for your answer...
I tried to use ctx but it does not work.....

For example : 

```
from buildbot.util import epoch2datetime

template = u'''\
 {{ MyDate }}
 {% set Epoch = buildset['submitted_at'] %}
'''

from buildbot.util import epoch2datetime
X = MyMessageFormatter(
   template = template,
   template_type = 'html',
   ctx = dict( MyDate=util.epoch2datetime(Epoch)),
   #ctx = dict( MyDate=datetime.fromtimestamp(Epoch).strftime('%c') ),
   subject = " {{ buildername }} : {{ statuses[results] }}",
   want_properties = True,
   want_steps = False,
   want_logs = True
)
```
Starting buildmaster raises a problem :

```
builtins.AttributeError: Unknown component name: epoch2datetime

Configuration Errors:
  error while parsing config file:  Unknown component name: epoch2datetime (traceback in logfile)
```

I'm sure I'm doing it wrong but I don't see where :-( 

any idea ?
Tkx

**Comment by jmbeuken** [comment, marked as answer]
I've got it!   :-)

For the record : 

```
class MyMessageFormatter(MessageFormatter):

    def buildAdditionalContext(self, master, ctx):
            Epoch = ctx['buildset']['submitted_at']
            ctx['MyDate'] =  datetime.fromtimestamp(Epoch).strftime('%c')
```
and in template :

`<tr><td> Time: </td><td><b>{{ MyDate }} </b></td></tr>`

A+
